internal/pkg/agent/storage/store/migrations.go

```diff
@@ -0,0 +1,201 @@
+// Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+// or more contributor license agreements. Licensed under the Elastic License;
+// you may not use this file except in compliance with the Elastic License.
+
+package store
+
+import (
+	"errors"
+	"fmt"
+	"time"
+
+	"github.com/elastic/elastic-agent/internal/pkg/agent/storage"
+	"github.com/elastic/elastic-agent/internal/pkg/agent/storage/store/internal/migrations"
+	"github.com/elastic/elastic-agent/internal/pkg/fleetapi"
+	"github.com/elastic/elastic-agent/pkg/core/logger"
+)
+
+var ErrInvalidYAML = errors.New("could not parse YAML")
+
+func migrateActionStoreToStateStore(
+	log *logger.Logger,
+	actionStorePath string,
+	stateDiskStore storage.Storage) (err error) {
+
+	log = log.Named("state_migration")
+	actionDiskStore, err := storage.NewDiskStore(actionStorePath)
+	if err != nil {
+		return fmt.Errorf(
+			"could not create disk store when migratins action store: %w", err)
+	}
+
+	stateStoreExists, err := stateDiskStore.Exists()
+	if err != nil {
+		log.Errorf("failed to check if state store exists: %v", err)
+		return err
+	}
+
+	// do not migrate if the state store already exists
+	if stateStoreExists {
+		log.Debugf("not attempting to migrare from action store: state store already exists")
+		return nil
+	}
+
+	actionStoreExists, err := actionDiskStore.Exists()
+	if err != nil {
+		log.Errorf("failed to check if action store %s exists: %v", actionStorePath, err)
+		return err
+	}
+
+	// nothing to migrate if the action store doesn't exist
+	if !actionStoreExists {
+		log.Debugf("action store %s doesn't exists, nothing to migrate", actionStorePath)
+		return nil
+	}
+	// delete the actions store file upon successful migration
+	defer func() {
+		if err == nil {
+			err = actionDiskStore.Delete()
+			if err != nil {
+				log.Errorf("failed to delete action store %s after migration: %v", actionStorePath, err)
+			}
+		}
+	}()
+
+	action, err := migrations.LoadActionStore(actionDiskStore)
+	if err != nil {
+		log.Errorf("failed to load action store for migration %s: %v",
+			actionStorePath, err)
+		return err
+	}
+
+	// no actions stored nothing to migrate
+	if action == nil {
+		log.Debugf("no action stored in the action store %s, nothing to migrate", actionStorePath)
+		return nil
+	}
+
+	if action.Type != fleetapi.ActionTypePolicyChange {
```

Comment:
Why do we ignore everything that isn't a policy change?

Reply:
Because there should not be any other type of action. IIRC the action store would silently reject any action that wasn't a policy change.
Also to avoid silently discarding actions, I [log a warning](https://github.com/elastic/elastic-agent/pull/4441/files#diff-8c3cf214bf832af7042942a6a48cd1a841471b1f99e3d4eb44c2e94b253f0267R80-R82) if it ever happens.

Reply:
I see it storing Unenroll as well. https://github.com/elastic/elastic-agent/blob/cb95c52546535d00fee6d753ea5c1ebee2c69e0d/internal/pkg/agent/storage/store/state_store.go#L169-L174

I am worried that someone will extend the store with a new action type in the future and forget to update the migrations to handle it. Is there a test we can write to detect this problem? Or should the migrations not be aware of action type restrictions at all?

Reply:
Or is this only handling actions that were persisted by previous versions of the agent which are known to only write this action type?

If that is the case, a link to the old code that proves this as a comment would help avoid this confusion later.